lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,506 @@
+use lightning::routing::network_graph;
+use lightning::util::ser::Readable;
+
+use crate::error::GraphSyncError;
+
+/// Backdating is used a mechanism to avoid accidentally overriding higher-fidelity gossip updates.
+/// Current delta amount is set to a week flat. Note that gossip updates are purged at the
+/// two-week-mark.
+///
+/// In the future, we will probably start delaying the purging of unsigned updates to their
+/// four-week-marks, or perhaps simply trust a provided server-timestamp that marks
+/// data receipt, as opposed to originally signed value. TODO: TBD
+const TIMESTAMP_RETROGRESSION_DELTA: u32 = 7 * 24 * 3600;
+
+/// Update network graph from binary data
+///
+/// `network_graph`: network graph to be updated
+///
+/// `update_data`: `&[u8]` binary stream that comprises the update data
+pub fn update_network_graph(
+	network_graph: &network_graph::NetworkGraph,
+	update_data: &[u8],
+) -> Result<(), GraphSyncError> {
+	let mut read_cursor = io::Cursor::new(update_data);
+	read_network_graph(&network_graph, &mut read_cursor)
+}
+
+pub(crate) fn read_network_graph<R: Read>(
+	network_graph: &network_graph::NetworkGraph,
+	read_cursor: &mut R,
+) -> Result<(), GraphSyncError> {
+	let mut prefix = Vec::with_capacity(4);
+	prefix.resize(4, 0);
+	read_cursor.read_exact(&mut prefix)?;
+
+	let _deserialize_unsigned = match &prefix[..] {
+		&[76, 68, 75, 1] => true,
+		_ => {
+			return Err(GraphSyncError::ProcessingError(
+				"Prefix must equal 'LDK' and the byte 2".to_string(),
+			));
+		}
+	};
+
+	// let chain_source: Option<Box<dyn lightning::chain::Access>> = None;
+	let current_timestamp = SystemTime::now()
+		.duration_since(UNIX_EPOCH)
+		.unwrap()
+		.as_secs() as u32;
+
+	let retrogressed_timestamp = current_timestamp - TIMESTAMP_RETROGRESSION_DELTA; // back-date updates by one week
+
+	// shadow variable for ownership release after length-restricted reads
+	let mut read_cursor = read_cursor;
+
+	let chain_hash: BlockHash = Readable::read(read_cursor)?;
+
+	let announcement_count: u32 = Readable::read(read_cursor)?;
+	for _ in 0..announcement_count {
+		let features = Readable::read(read_cursor)?;
+		let short_channel_id = Readable::read(read_cursor)?;
+		let node_id_1 = Readable::read(read_cursor)?;
+		let node_id_2 = Readable::read(read_cursor)?;
+
+		let announcement_result = network_graph.update_channel_from_announcement_fragments(
+			short_channel_id,
+			retrogressed_timestamp as u64,
+			features,
+			node_id_1,
+			node_id_2
+		);
+		if let Err(lightning_error) = announcement_result {
+			if let ErrorAction::IgnoreDuplicateGossip = lightning_error.action {
+				// everything is fine, just a duplicate channel announcement
+			} else {
+				return Err(lightning_error.into());
+			}
+		}
+	}
+
+	let update_count: u32 = Readable::read(read_cursor)?;
+	for _current_update_index in 0..update_count {
+		// println!("update index: {}", _current_update_index);
+
+		let short_channel_id: u64 = Readable::read(read_cursor)?;
```

Comment:
Same here, maybe sort them and put a difference instead of the full one?

Reply:
done, using BigSize